source/common/ssl/connection_impl.h

```diff
@@ -34,25 +28,44 @@ class ConnectionImpl : public Network::ConnectionImpl, public Connection {
   std::string subjectPeerCertificate() const override;
   std::string uriSanPeerCertificate() override;
 
+  // Network::TransportSocket
+  void setTransportSocketCallbacks(Network::TransportSocketCallbacks& callbacks) override;
+  std::string protocol() const override;
+  bool canFlushClose() override { return handshake_complete_; }
+  void closeSocket(Network::ConnectionEvent close_type) override;
+  Network::IoResult doRead(Buffer::Instance& read_buffer) override;
+  Network::IoResult doWrite(Buffer::Instance& write_buffer) override;
+  void onConnected() override;
+
   SSL* rawSslForTest() { return ssl_.get(); }
 
 private:
-  PostIoAction doHandshake();
+  Network::PostIoAction doHandshake();
   void drainErrorQueue();
   std::string getUriSanFromCertificate(X509* cert);
 
-  // Network::ConnectionImpl
-  bool canFlushClose() override { return handshake_complete_; }
-  void closeSocket(Network::ConnectionEvent close_type) override;
-  IoResult doReadFromSocket() override;
-  IoResult doWriteToSocket() override;
-  void onConnected() override;
-
+  Network::TransportSocketCallbacks* callbacks_{};
   ContextImpl& ctx_;
   bssl::UniquePtr<SSL> ssl_;
   bool handshake_complete_{};
 };
 
+class ConnectionImpl : public Network::ConnectionImpl {
+public:
+  ConnectionImpl(Event::DispatcherImpl& dispatcher, int fd,
+                 Network::Address::InstanceConstSharedPtr remote_address,
+                 Network::Address::InstanceConstSharedPtr local_address,
+                 Network::Address::InstanceConstSharedPtr bind_to_address, bool using_original_dst,
+                 bool connected, Context& ctx, InitialState state);
+  ~ConnectionImpl();
+
+  // Network::Connection
+  Ssl::Connection* ssl() override { return dynamic_cast<SslSocket*>(transport_socket_.get()); }
```

Comment:
Hm.  If ssl() is a weird function on the base connection class (which returns nullptr) what do you think of making it an additional weird function on the transport socket?  It'd let us avoid the dynamic cast, and I think do away with the ConnectionImpl entirely (as we could create the base connection class with a new SslSocket explicitly, and the regular connection class would call the null-returning ssl() for the raw socket.

Reply:
Yes, I'm planning to remove the whole Ssl::ConnectionImpl at the end the ssl() is kept here for compatibility with others, as it is used widely in filters, etc. That refactor will involve many classes, can be done in a separate PR.

Reply:
Cool, let's TODO that cleanup just so it's clear what the plan is.  I'm fine with a follow-up to keep this clean and simple

Reply:
Done